### ISSUE TITLE
Failed to register: StatusCode: 429 (RegistrationRetryException)

### ISSUE THREAD

**Issue by Cangoo** [bbernhard/signal-cli-rest-api#441]
### The problem

Get {"error":"Failed to register: StatusCode: 429 (RegistrationRetryException)\n" when trying to register number with catpure.

### Are you using the latest released version?

- [X] Yes

### Have you read the troubleshooting page?

- [X] Yes

### What type of installation are you running?

signal-cli-rest-api Docker Container

### In which mode are you using the docker container?

Normal Mode

### What's the architecture of your host system?

arm64

### Additional information

I'm trying to set up the docker container for homeassistant for a land-line number using this readme:
https://github.com/bbernhard/signal-cli-rest-api/blob/master/doc/HOMEASSISTANT.md

Using this line of code:
`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": true}' 'http://127.0.0.1:8080/v1/register/+43123456789`

I got the reply:
`{"error":"Captcha required for verification, use --captcha CAPTCHA\nTo get the token, go to https://signalcaptchas.org/registration/generate.html\nAfter solving the captcha, right-click on the \"Open Signal\" link and copy the link.\n"}`

So I went to https://signalcaptchas.org/registration/generate, solved the capture and got a link like:
signalcaptcha://signal-hcaptcha.5fad9...aEIVfjTLs4E

As descripted in the readme, I then did:
`curl -X POST -H "Content-Type: application/json" -d '{"captcha":"signal-hcaptcha.5fad9...aEIVfjTLs4E", "use_voice": true}' 'http://127.0.0.1:8080/v1/register/+43123456789'`

But I only get the reply:
`{"error":"Failed to register: StatusCode: 429 (RegistrationRetryException)\n"`

I couldn't find any information about this error. The land-line number I use wasn't registered for Signal.

Any idea what might be wrong?

**Comment by Cangoo** [comment]
I found out that when deleting all files in /signal-cli-config/, the error "RegistrationRetryException" vanishes. 

Can't check if it fully works because I get a "Rate limit exceeded"-error now. So waiting for that to vanish too...

**Comment by Cangoo** [comment]
Nope, still get {"error":"Failed to register: StatusCode: 429 (RegistrationRetryException)\n". Any ideas?

**Comment by Cangoo** [comment]
Hi and thanks for your reply and the hint with the signalcaptcha:// - very helpful.

I now managed to get the call for the verification number by doing this (see https://github.com/AsamK/signal-cli/wiki/Quickstart#set-up-an-account):

`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": false, "captcha":"signalcaptcha://signal-hcaptcha.5fad...Rk--Ro"}' 'http://127.0.0.1:8080/v1/register/+43123456789'`

**WAIT 60 sec**

`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": true, "captcha":"signalcaptcha://signal-hcaptcha.5fad...YmYTv56fA"}' 'http://127.0.0.1:8080/v1/register/+43123456789'`

The link above claims that Failed to register: [429] Rate limit exceeded: 429 (RateLimitException) might be an error if you are not doing this correctly (first use voice=false, then wait, then voice=true).

But now I'm stuck at the same point as you:
`curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:8080/v1/register/+43123456789/verify/123-456'`
=> {"error":"Verify error: StatusCode: 404\n"}

Do you now if the format for the code is '123-456' or '123456'? Okay, both didn't work...

**Comment by Cangoo** [comment]
Uh, oh, eh, although I got the verification-error, I'm now able to send messages. Yay!

**Comment by Cangoo** [comment]
Not sure what did the trick. I tried:

`curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:8080/v1/register/+43123456789/verify/123-456'`

and 

`curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:8080/v1/register/+43123456789/verify/123456'`

several times, I even called: http://127.0.0.1:8080/v1/register/+43123456789/verify/123-456 in the browser, always getting "Verify error: StatusCode: 404". I then out of fun tried to receive messages which suddenly worked.

**Comment by mwllgr** [comment]
Hey, got it working by working 60 seconds before setting `use_voice` to `true` too. I also removed the hyphen in the registration code in the verify URL.

**Comment by Cangoo** [comment]
Hello, I just tried it with another landline-number and it worked like charm on the first try using no hyphen in the code to verify the number.

**Comment by flocsi** [comment]
Does not work for me. Verifying results in "StatusCode: 404"

**Comment by bbernhard** [comment]
please try with Version 0.80

**Comment by flocsi** [comment]
Even with V 0.80, I receive "Verify error: StatusCode: 404" after entering the code I received via phone call.

**Comment by alexyao2015** [comment]
I found that I was getting 429 when I specified `use_voice` to true initially. It seems that you just set `use_voice` to false, make a request, then set `use_voice` to true. And only provide it a captcha after getting an error stating it is required. I also found that on my second attempt after setting `use_voice` to true, it required another request with the same payload to give me a successful response.

For me, the following worked:
1. `use_voice` false without captcha
2. `use_voice` false with captcha
3. `use_voice` true without captcha ---> 429
4. `use_voice` true without captcha ---> 201

**Comment by flocsi** [comment]
I get "StatusCode: 502 (ExternalServiceFailureException)". Perhaps, it is an issue with this particular landline number that has been used earlier.

**Comment by PeterNSteinmetz** [comment]
Running on the machine running the signal-cli-rest-api I have tried the following:

`curl -X POST -H "Content-Type: application/json" --data "{\"use_voice\": false}" 'http://10.0.0.44:8080/v1/register/+14807751111'`
which gives the '{"error":"Captcha required for verification, '

I then generate a captcha on my laptop and try 

`url -X POST -H "Content-Type: application/json" --data "{\"use_voice\": false, \"captcha\":\"signalcaptcha://s
ignal-hcaptcha.5fad97ac-7d06-4e44-b18a- ... "}" 'http://1
0.0.0.44:8080/v1/register/+14807751111'`

which gives '{"error":"Invalid captcha given.\n"}' 

If I try the same command again, it gives '{"error":"Failed to register: StatusCode: 429 (RegistrationRetryException)\n"'

Does the captcha need to be generated on the same machine that the rest api is running on? 
And then the 429 is generated because of retrying too quickly?

**Comment by bbernhard** [comment]
The actual machine doesn't matter. But you need to be fast, as those captcha tokens are really short-lived.

**Comment by PeterNSteinmetz** [comment]
Thanks. So the first request should be with use_voice false and with a captcha? Then the second request after 60 seconds with use_voice=true should also have a captcha which is different?

**Comment by bbernhard** [comment]
> Thanks. So the first request should be with use_voice false and with a captcha? Then the second request after 60 seconds with use_voice=true should also have a captcha which is different?
> 
> How would I remove those files when the server is running inside the docker container?

I never needed to do the dance that some people needed to do in order to get a number registered. I think the Signal Server has good days and bad days and on those bad days the registration just doesn't work (I guess they have some sort of abuse protection in place, which might accidentally trigger). 

So, ideally it should just work, no matter if you use `use_voice` or not. As already mentioned, the captcha token that is returned is really short lived, so you need to be fast. 

> How would I remove those files when the server is running inside the docker container?

Did you bind mount the folder from the host system to the docker container? If so, just remove the all the files inside and restart the container. 

If you instead use docker volumes, you can find the docker container id of the running docker container with `docker ps` and then connect to the running docker container with `docker exec -it <container id> /bin/bash`. Inside the docker container remove everything inside the `/home/.local/share/signal-cli/` folder. Afterwards, restart the docker container and try again.

**Comment by PeterNSteinmetz** [comment]
@bbernhard Thanks again for the suggestions and your work on this rest API. 

I am running this under HomeAssistant so using the docker container at ghcr.io/haberda/signal_messenger/aarch64:0.88.0 . By stopping the add-on in HA the docker container went away and then starting it again presumably cleared the configurations files.

The following then worked for me:

`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": false, "captcha":"signalcaptcha://signal-hcaptcha.5fad97ac-7d06-4e44-b18a- ... "}' 'http://10.0.0.44:8080/v1/register/+14807751111'`

Wait one minute.

`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": true}' 'http://10.0.0.44:8080/v1/register/+14807751111'`

That phone number then rang and it gave me the verification code by voice.

`curl -X POST -H "Content-Type: application/json" 'http://10.0.0.44:8080/v1/register/+14807751111/verify/123456'`

I was then able to send a message with
`curl -X POST -H "Content-Type: application/json" --data '{"message":"Test from HA.","number":"+14807751111","recipients":["+14808622222"]}' 'http://10.0.0.44:8080/v2/send'`

None of these commands returned an error code or output. The captcha itself was simply generated on another machine that has the same external ip as the machine running homeassistant.

I pre-composed the commands in a text editor so I could rapidly copy the captcha link into the command and then execute it.

**Comment by enaut** [comment]
How long is "short lived" like 10 sec or 60+ sec?

**Comment by PeterNSteinmetz** [comment]
For it seemed like it needed to be within 5 seconds or so.

**Comment by jsschmid** [comment]
I'm having kind of the same problem, but from error 429 I'm down to
`"error":"Failed to register: StatusCode: 502 (ExternalServiceFailureException)"`

I did the following
1. Issuing a register without captcha
`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": false}' 'http://127.0.0.1:8088/v1/register/+493xxxxxxxxx'`
2. Issuing the same request with captcha as demanded ... I posted this withing 10 seconds or so.
`curl -X POST -H "Content-Type: application/json" --data '{"use_voice": false, "captcha":"signalcaptcha://signal-hcaptcha.5fad97ac-7d06-4e44-b18a-b950b20148ff.registration.P1_eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.hadwYXNza2V5xQUaOrv1sr8NKGywlY3D8z8bcWZ5RuQLoi5i0M1pWzEq4GLAXeOZXPcEkqLGjp5jB4uN9GCBHSbgvuTAbeZ6CYX30AVAvf0-eli5q9qJhrEzrh1BOjlpxARvqn30XUZPmzHoNnl5evxlUcIeX5GmpC0RQcjusRY4jX23WhEya6Je9fuq5ryR_DWuppuETtR1O9pbH0YhaeQDoJaU4qGKNI5TxWf127U_R_O_mw1FBuk7yVXpn4Mhpld1QBOWtAYUEwAtDGd06f6TsZvs1Ljd3MVOJt94wStybp3LfUzRR_ZJ8LRPNUZ2-HObrZEqJ4YgvzAUnTlgR371UuucM4kC8nengKKdnQeHLLdGeG9FrA7xyXFN3t3Md11G-Wp9CrrHONgx7YLBfXOwYFhkeqjxk1gwRFbBalRx5SRcldU9zy63tPYPISiRv1x3Xx3DCAGHgbyVbX0_-qorSRPyLS3CgM1cGXJjNBq0gjdtCJsmJs_fEvqB0tfZNEty2EJ-_cEJOoVKh655jsI8v8q2yAGPV-MWN1M3vNi6KqZWhW_QEnYTjKXceIYkYyRnuWEjmBtF7nHdxM7m4gOyKCmESt0AZz5a8JXFt2T2CJiJNk-DUSkQxMFq4Xz5z9eYRePCW101Tg3at4zVcts5FFkgpzxrvcrqJSYRdU5IHWF8ufqeesd8qoPU-ZJ2kTOh2e8AHDNWAp9sJA33-6i3aFQCcKCBxoRYmGRrJV_EiWpygBZyxEdXM9EtclkyjeBhJaJjInduZrSPXdBgCYyKxH6_HUkbYV70ICSCWZ0czT7K6vcNr2zUSzqqbDdD0ZUP8EGdCheN2MQ0zqH_ymAaV-0pKVDQBjdcsZCzOcIWGStb2a3u4TghKMLDkutpjvSU7yMglvey96FScKUNvtdJzHwXqautTi-R2dOdtuinN3T08PtP38184pPp6mmz08pfiUhzBQ9XsSydwqcWlXXTxWgAyRD-K1nZgrZRULF3wlw0njdXAxCVY1_URYTqIHIoKfuyD5rUC5GglUlKLUBG_bscMgnS0TXari53TYvEEc0aLrNK_Vki6BPn4rEeO9mZ5RK8cNdRZY3umUBj9gDtUIrf9bxXPyJ97QDaaCg7ryzi37W1wKnYdk-vRSCM3NFmB9dmpdxBznkKccm9dEIYzkkGowyJSKPJqNhA0ppDaAHaKyTR8OiAiCdpUvDlOWqpRh_3ouHADGM91xOHcLpTOzzrIlupSckAawYGyPRzluHOgv47eZOD9MBF7tzh46_s_7W14jPDX-HUePGjo0HgA_E-ogstKglVMYOjNVZnnaVR_FWQS57xKLZnQ8pqxw099kzzSNbcsLXR_l4q0cQQITCCJWjXtNAvmYn8bCdbarCW2_4Ru-OUmDzulQAW5PWQJng7XsSGLBjNr7Lnj8nsceoYzhB9OfoxWw35GLysPGR8uarPEzYn2xGWkSrKq50OJiJUGcTWHG1hDCIy-xekHDn_Qbs9e5cxm2_Pt1i0mkCMJc92bZlfynVFjusMdvn2oUR9TZc3v50iKf5lMsu7guvnGhAFqPyRrnsnUAncvsngQ4Th3d7TWqtt6nJRXzi8HJk8loxwVqB9SveWqtz50ERkrM0Zqzj6Y8BGaWO5ULpL3RyJq3HNwmouNbAZqOOfdCC7HDwij_hvsKy-j9L5l2qubJnHnWcqI1_qOTH30tedINPV_QAQsoS-o79SN8OkkJwE1NAtIF4wDp1RVtyfHNSaOaNleHDOZ3Zt4qhzaGFyZF9pZM4UPIQfomtyqDMwM2RlZDZionBkAA.ITfRKFe58NTaOpHlmbm4fcBKtIVJ-opxBDJ_rdFLJ2w"}' 'http://127.0.0.1:8088/v1/register/+493xxxxxxxxx'`
3. ...got the error, began to cry silently as I tried this about 7 times yesterday already.
`{"error":"Failed to register: StatusCode: 502 (ExternalServiceFailureException)"}`

Please help !

**Comment by enaut** [comment]
With landline for me it was the same (external service exception). The next call (with voice true and a new captcha less than 10sec old) worked then... But I had to wait more than 24h for the rate limit to expire. I tried a lot though so I'm not actually sure what got me through I think it was:
  1. wait for a day or two
  2. "use_voice": false -> 502
  3. immediately after "use_voice": true with new captcha -> success

**Comment by SprezzaturaV** [comment]
After some playing around and adding my partner's number to a group before they registered their number on signal. The number I'd wanted to use got kind of flagged and I got an Error in HA. I had successfully sent messages before that. When I tried to send a test message via terminal: 

`curl -X POST -H "Content-Type: application/json" -d '{"message": "Hello World!", "number": "+4311111111", "recipients": ["+4322222222"]}' 'http://192.x.x.x:8080/v2/send'`

I got an error limit exceeded error 429 with a the signal captcha generator [link]([https://signalcaptchas.org/registration/generate](url)) and also a challenge token. I managed to challenge the limit with following command syntax via terminal in HA:

`curl -X POST -H "Content-Type: application/json" -d '{"captcha":"signalcaptcha://signal-hcaptcha.5fad97ac-7d06-4e44-b18a-b950b20148ff.challenge.P1_eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJwYXNza2V5Ijoiay8rRWpxakViM1hwKzRmNy90VUM1RmZ6ZHJPeVppUUNreERZVkxxYlBNUHB5RzJwMUpDa01aT29FUUE4L3ZIQ09CVHowd0JnajFLcTVqazk0dm54b3h3V2RmVWZ6aVZrMnZOZFQrWnNoa1NhTUtPL0l5Y0w3alhReCtudVZtOHVhdWZFVmVEQ2ZrWXB4bmZvWHk3WU11RHNJMjMwZHZwazFjeUdMWHBYVytRb25iUXc2M0Y1SExONDZqZW5PWHJjTUI1bDhCWHFLTVdpQUJWZDJDWGE5MUdZOXVVdmZZUWlHbldZcldmYVZRcStyUVVCb2xrNnJkMy9sc09yVVFMZ0dMNWQ5M1pacTJiTWJoVTI5K1V0SU0zNG1lTjB2cUxZSXJpVW9BdU1RZkRNclVkalNyR2tHQjBocTNzUFcrdko2NStwY1dkck10WFduNG8rbUZzd0dYV29yQUR6WHp1d2l5a0toVWQvRjRsQWdVMFgzay91U1lzbDhVSlpTYUliYjI0ZmdNM3VFbWRxcHRYSE9pQktKVEtyZUI2NC9oMFRCSENHR0pMZCtueXBhcmRVSXlRcjM4bmlNT2doSmNQZmZjQzgyZ0ZjV1hoTmxNMVJZUXVic0Jkb1FqYk1MbEl4RWdpL0pFVTFXdVlyKy9OQ1BPUXFVMDkvYmxsUDdVd3lEYXJvSjZrUFZQaTkyMGQ4TnllZkNYYW51TTQwYkFZak9oRnVKZWNWcEtScHR5MGQwS3dyYldFS0YzRHg3TDVyYzM4U25reFZiM051M2N1VGYreEpDWEh1QlBxV3dRNE5KcmFQSk5qbENXVm41c0lCc0szSGxaVnY4ZlFWd3BDbjUwY1Z5dUhLNjB6bUhEcGlaNldHU2ZGcmh3ZG5OTXBSa1dlcnNLYytjNnZyb3dLREptVE9aNURwMzc2YWo0cHVSSzhUaGx1N2RYYTI2MG16K0o5ckp5OWFQQ3RVbk9KQmJGQkJmcy9MU09wbmZLNTAwYktNLys4cUZBVEMrR0U1bEtQaWdjTkNQQnEzOEs1SzNpSlcyTVI1SitLT2E0T0J3d2h1cmJZUVdUNnE0WGpoRmZkUkg1R0toQUdIV2lPT0Z3Vjh2Q0lFV1dLSmNWTXhhY0JwQXlNcFZZZ29LU3p2Y2lMMHM0SlQ1UEpyNWVvRDBkeVQvcEFpYUVka3BSbDlCd3FIUnRDaFhxxxxxxxxMWZNWkxheXB4YVpYOVIxdHVLM05yZG16eElEa3cybVZWYzk0eURyN3pzcEsxV1g2M2FObEN2MW9kbHlZelkxUThrMHVSZjBscU5qN3RBY1dBVzNvVTVLY2lXeUo1YjBuNldneEwxUnM0Ukw5algvOTRGdTE1c01YVituUi9xenhMOUovQldsd2xZS3JyUTdEdmpPcjM0czkzdXdhMUY0czVIZlloNkU4dWlLcm02ekNSZ2RQVE1MZ2dENUZYTUxsY3M3cjJVL3NQajVDd3JIaUVBTVNuN3JJd3JveUdlUXlvWGNINWd3YWJnSzNwMkZPN3F1M25oMXgyTnZSMGs2SUtNbGtNdXFpajNCN0VuL2YxL1J5QTUycFdWV1JOdlJubkNCYStrd0RTaDVMT2N2bmdveHZGOWs4ZWtORWJLTkhGbXRLR29RLzUzY2lMYm10TE9QM2NhdmtSUmZuODJrWUgxanlHYkZhNTdVVEM3b1FCdm44ZDliWmZvblhOSmpNYkFjRHFmdTVVN3pyKy8vdjlUVklYcnhzZm03OXUzbzR2eHdNZ0REbng4aEh6VEhZQ2NtRUNTVXBpVjNLajNlM0pSam9nNzIrM1VqbG83clZPbTJQeXM3WnVZQlRxxxxxxxxxxxxxxxxxxAxWlJZNTBuMVRCM1JpdFFML3k3Vk01QnNET2JpdVRpTEhFeUJPeStRZWdqUkdzb1ZPODBpMzM2MUVRMDYvYUlhdVgwdkNJNmxRQlVDMUdRNE04OWtNQWVFUE9DSlhheGtZSllZdWtHcEFvSW96cEh2TThFdkVCcjRFc0Y4dXNnamxZZ01XYkcrUFoyYkYvdTlrY2gyT3dYL29INEswb2dqQ1kwcmQwYzNSWjF6VWdaNUpCOFY3Y3hwMk5najVWMElxenI1OWpCdzFxM1FMQnNNdWp6WmVkVmlXVmUrVnZ4TG02WGJnNzduaEZzb3kvNWs4bEkvVDh0Z0g5L3lMQXVCeDZwL2tzRFJvL3BtaUVEMVRCenJPVUxxxxxxxxxxm9zV2pBejJXaGl4UDA5MTcrVnpkRGN6Qm9ZQVl6QlpyUnJRaitMZDVsd1MzclhMR2gvZUJ4Mi9hQVVmanN0cm8ra3Y0eVppTUZmZUNkeGZ1a09NTTRuRzRZN21lMzlraE91R2JKbGkxMFdPMi95b0EwemRMTEQyNFJzS2VEYmZHVDZQS012dTFSamN2UG80bUd3TSsxZjdJTlVBb3F1RFM5TW5JM0pPWmpkeHM5bVVneHAzMlZiODgzeSs5IiwiZXhwIjoxNzQxMzQ4MDQ2LCJzaGFyZF9pZCI6MzM5NTEwMzAzLCJrciI6IjMxMWY5ZmRmIiwicGQiOjB9.9ffGrjsK4A9rO-ntWMt9iMvMBm1ssw9YvftiSzizisY", "challenge_token":"499e57f6-xxxx-81ee-bac9d07fb9fe"}' 'http://192.x.x.x:8080/v1/accounts/+4311111111/rate-limit-challenge'`
